### PR TITLE
3.2 remove node property descriptor

### PR DESCRIPTION
**Pull request by ragadeeshu**

### CODE REVIEW

File: community/kernel/src/main/java/org/neo4j/kernel/api/schema/EntityPropertyDescriptor.java [hunk 1]
```diff
@@ -77,15 +77,15 @@ public int getEntityId()
     /**
      * @return property key token id this index is for.
      */
-    public int getPropertyKeyId()
+    public int getPropertyId()
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/api/schema/EntityPropertyDescriptor.java:80]: This class (`EntityPropertyDescriptor`) is going away soon too, right?

File: community/kernel/src/main/java/org/neo4j/kernel/api/schema_new/LabelSchemaDescriptor.java [hunk 2]
```diff
@@ -28,7 +28,7 @@
     private final int labelId;
     private final int[] propertyIds;
 
-    LabelSchemaDescriptor( int labelId, int... propertyIds )
+    public LabelSchemaDescriptor( int labelId, int... propertyIds )
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/api/schema_new/LabelSchemaDescriptor.java]: This should not be public, we only construct via the factory.

File: community/kernel/src/main/java/org/neo4j/kernel/api/schema_new/constaints/ConstraintBoundary.java [hunk 3]
```diff
@@ -108,12 +108,14 @@ public PropertyConstraint computeSpecific( LabelSchemaDescriptor schema )
             switch ( descriptor.type() )
             {
             case UNIQUE:
-                return new UniquenessConstraint(  new NodePropertyDescriptor(
-                        schema.getLabelId(), schema.getPropertyId() ) );
+                final int labelId = schema.getLabelId();
+                final int[] propertyKeyIds = new int[]{schema.getPropertyId()};
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/api/schema_new/constaints/ConstraintBoundary.java]: You can just do `schema.getPropertyIds()` to get the real array.

File: community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/IndexProcedures.java [hunk 4]
```diff
@@ -63,7 +63,7 @@ public void resampleIndex( String indexSpecification ) throws ProcedureException
     {
         IndexSpecifier index = parse( indexSpecification );
         int labelId = getLabelId( index.label() );
-        int propertyKeyId = getPropertyKeyId( index.property() );
+        int propertyKeyId = getPropertyId( index.property() );
         //TODO: Support composite indexes
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/IndexProcedures.java:58]: Made card for fixing these TODOs

File: community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/IndexProcedures.java [hunk 5]
```diff
@@ -111,7 +111,9 @@ private NewIndexDescriptor getIndex( int labelId, int propertyKeyId, IndexSpecif
     {
         try
         {
-            return operations.indexGetForLabelAndPropertyKey( new NodePropertyDescriptor( labelId, propertyKeyId ) );
+            final int labelId1 = labelId;
+            final int[] propertyKeyIds = new int[]{propertyKeyId};
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/IndexProcedures.java]: `getPropertyIds()` as above

File: community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java [hunk 6]
```diff
@@ -711,12 +709,7 @@ public void constraintDrop( KernelStatement state, ConstraintDescriptor constrai
     }
 
     @Override
-    public NewIndexDescriptor indexGetForLabelAndPropertyKey( KernelStatement state, NodePropertyDescriptor descriptor )
-    {
-        return indexGetForLabelAndPropertyKey( state, SchemaBoundary.map( descriptor ) );
-    }
-
-    private NewIndexDescriptor indexGetForLabelAndPropertyKey( KernelStatement state, LabelSchemaDescriptor descriptor )
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/api/StateHandlingStatementOperations.java:719]: Nice =)

File: community/kernel/src/main/java/org/neo4j/kernel/impl/coreapi/schema/PropertyNameUtils.java [hunk 7]
```diff
@@ -28,19 +28,19 @@
...
     {
     }
 
-    public static String[] getPropertyKeys( ReadOperations readOperations, NodePropertyDescriptor descriptor )
+    public static String[] getPropertyKeys( ReadOperations readOperations, LabelSchemaDescriptor descriptor )
             throws PropertyKeyIdNotFoundKernelException
     {
         int[] propertyKeyIds =
-                descriptor.isComposite() ? descriptor.getPropertyKeyIds() : new int[]{descriptor.getPropertyKeyId()};
+                false ? descriptor.getPropertyIds() : new int[]{descriptor.getPropertyId()};
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/coreapi/schema/PropertyNameUtils.java]: Whaat?

File: community/kernel/src/main/java/org/neo4j/kernel/impl/coreapi/schema/PropertyNameUtils.java [hunk 8]
```diff
@@ -49,10 +49,10 @@ private PropertyNameUtils()
         return propertyKeys;
     }
 
-    public static String[] getPropertyKeys( TokenNameLookup tokenNameLookup, NodePropertyDescriptor descriptor )
+    public static String[] getPropertyKeys( TokenNameLookup tokenNameLookup, LabelSchemaDescriptor descriptor )
     {
         int[] propertyKeyIds =
-                descriptor.isComposite() ? descriptor.getPropertyKeyIds() : new int[]{descriptor.getPropertyKeyId()};
+                false ? descriptor.getPropertyIds() : new int[]{descriptor.getPropertyId()};
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/coreapi/schema/PropertyNameUtils.java]: Whaat 2

File: community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java [hunk 9]
```diff
@@ -615,8 +616,10 @@ private NewIndexDescriptor findAnyIndexByLabelAndProperty( ReadOperations readOp
     {
         try
         {
+            final int labelId1 = labelId;
+            final int[] propertyKeyIds = new int[]{propertyId};
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/factory/GraphDatabaseFacade.java]: `getPropertyIds`

File: community/kernel/src/main/java/org/neo4j/kernel/impl/util/dbstructure/DbStructureArgumentFormatter.java [hunk 10]
```diff
@@ -98,32 +100,35 @@ else if ( arg instanceof NewIndexDescriptor )
...
-                    constraint.descriptor().propertyIdText() ) );
+            builder.append( format( "new UniquenessConstraint( SchemaDescriptorFactory.forLabel( %d, %s ) )", labelId,
+                    asString( constraint.descriptor().getPropertyIds() ) ) );
         }
         else if ( arg instanceof NodePropertyExistenceConstraint )
         {
             NodePropertyExistenceConstraint constraint = (NodePropertyExistenceConstraint) arg;
             int labelId = constraint.label();
             builder.append(
                     format( "new NodePropertyExistenceConstraint( new NodePropertyDescriptor( %s, %s ) )", labelId,
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/util/dbstructure/DbStructureArgumentFormatter.java]: Missed `NodePropertyDescriptor`

File: community/kernel/src/main/java/org/neo4j/kernel/impl/util/dbstructure/DbStructureArgumentFormatter.java [hunk 11]
```diff
@@ -98,32 +100,35 @@ else if ( arg instanceof NewIndexDescriptor )
...
-            RelationshipPropertyDescriptor descriptor = ((RelationshipPropertyExistenceConstraint) arg).descriptor();
-            int relTypeId = descriptor.getRelationshipTypeId();
-            int propertyKey = descriptor.getPropertyKeyId();
+            RelationTypeSchemaDescriptor descriptor = (RelationTypeSchemaDescriptor) (
+                    (RelationshipPropertyExistenceConstraint) arg)
+                    .descriptor();
+            int relTypeId = descriptor.getRelTypeId();
+            int propertyKey = descriptor.getPropertyId();
             builder.append(
                     format( "new RelationshipPropertyExistenceConstraint( new NodePropertyDescriptor( %s, %s ) )",
```
Comment (fickludd) [inline community/kernel/src/main/java/org/neo4j/kernel/impl/util/dbstructure/DbStructureArgumentFormatter.java]: Missed `NodePropertyDescriptor`

File: community/kernel/src/test/java/org/neo4j/kernel/builtinprocs/ResampleIndexProcedureTest.java [hunk 12]
```diff
@@ -94,7 +94,8 @@ public void shouldLookUpTheIndexByLabelIdAndPropertyKeyId()
 
         procedure.resampleIndex( ":Person(name)" );
 
-        verify( operations ).indexGetForLabelAndPropertyKey( new NodePropertyDescriptor( 123, 456 ) );
+        final int[] propertyKeyIds = new int[]{456};
```
Comment (fickludd) [inline community/kernel/src/test/java/org/neo4j/kernel/builtinprocs/ResampleIndexProcedureTest.java]: `getPropertyIds`

File: community/kernel/src/test/java/org/neo4j/kernel/impl/api/state/SchemaTransactionStateTest.java [hunk 13]
```diff
@@ -67,7 +67,9 @@ private static NewIndexDescriptor indexCreate( StateHandlingStatementOperations
     private static NewIndexDescriptor indexGetForLabelAndPropertyKey(
             StateHandlingStatementOperations txContext, KernelStatement state, int labelId, int propertyKey )
     {
-        return txContext.indexGetForLabelAndPropertyKey( state, new NodePropertyDescriptor( labelId, propertyKey ) );
+        final int labelId3 = labelId;
+        final int[] propertyKeyIds = new int[]{propertyKey};
```
Comment (fickludd) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/api/state/SchemaTransactionStateTest.java]: ping

File: community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/command/Commands.java [hunk 14]
```diff
@@ -141,11 +140,11 @@ public static RelationshipGroupCommand createRelationshipGroup( long id, int typ
     }
 
     public static SchemaRuleCommand createIndexRule( SchemaIndexProvider.Descriptor provider,
-            long id, NodePropertyDescriptor descriptor )
+            long id, LabelSchemaDescriptor descriptor )
     {
         SchemaRule rule = IndexRule.indexRule(
                 id,
-                NewIndexDescriptorFactory.forSchema( SchemaBoundary.map( descriptor ) ),
+                NewIndexDescriptorFactory.forLabel( descriptor.getLabelId(), descriptor.getPropertyId() ),
```
Comment (fickludd) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/command/Commands.java]: Can still be `forSchema`

File: community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/command/HighIdTransactionApplierTest.java [hunk 15]
```diff
@@ -74,10 +75,12 @@ public void shouldUpdateHighIdsOnExternalTransaction() throws Exception
...
 
         // Schema rules
+        final int[] propertyKeyIds = new int[]{1};
         tracker.visitSchemaRuleCommand( Commands.createIndexRule(
-                NO_INDEX_PROVIDER.getProviderDescriptor(), 10, new NodePropertyDescriptor( 0, 1 ) ) );
+                NO_INDEX_PROVIDER.getProviderDescriptor(), 10, SchemaDescriptorFactory.forLabel( 0, propertyKeyIds ) ) );
+        final int[] propertyKeyIds1 = new int[]{2};
         tracker.visitSchemaRuleCommand( Commands.createIndexRule(
-                NO_INDEX_PROVIDER.getProviderDescriptor(), 20, new NodePropertyDescriptor( 1, 2 ) ) );
+                NO_INDEX_PROVIDER.getProviderDescriptor(), 20, SchemaDescriptorFactory.forLabel( 1, propertyKeyIds1 ) ) );
```
Comment (fickludd) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/transaction/command/HighIdTransactionApplierTest.java]: `getPropertyIds`

File: community/kernel/src/test/java/org/neo4j/kernel/impl/util/dbstructure/DbStructureArgumentFormatterTest.java [hunk 16]
```diff
@@ -72,8 +73,9 @@ public void shouldFormatIndexDescriptors()
     @Test
     public void shouldFormatUniquenessConstraints()
     {
-        assertEquals( "new UniquenessConstraint( new NodePropertyDescriptor( 23, 42 ) )",
-                formatArgument( new UniquenessConstraint( new NodePropertyDescriptor( 23, 42 ) ) ) );
+        final int[] propertyKeyIds = new int[]{42};
```
Comment (fickludd) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/util/dbstructure/DbStructureArgumentFormatterTest.java]: not needed

File: community/kernel/src/test/java/org/neo4j/kernel/impl/util/dbstructure/DbStructureCollectorTest.java [hunk 17]
```diff
@@ -44,7 +44,8 @@ public void collectsDbStructure()
         collector.visitRelationshipType( 1, "LIVES_IN" );
         collector.visitRelationshipType( 2, "FRIEND" );
         collector.visitUniqueIndex( NewIndexDescriptorFactory.forLabel( 1, 1 ), ":Person(name)", 1.0d, 1L );
-        collector.visitUniqueConstraint( new UniquenessConstraint( new NodePropertyDescriptor( 2, 1 ) ), ":Person(name)" );
+        final int[] propertyKeyIds = new int[]{1};
```
Comment (fickludd) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/util/dbstructure/DbStructureCollectorTest.java]: ping

File: community/kernel/src/test/java/org/neo4j/kernel/impl/util/dbstructure/DbStructureInvocationTracingAcceptanceTest.java [hunk 18]
```diff
@@ -133,7 +134,9 @@ private void exerciseVisitor( Function<Object, DbStructureVisitor> visitor )
         visitor.apply( null ).visitRelationshipType( 1, "REJECTS" );
         visitor.apply( null ).visitIndex( NewIndexDescriptorFactory.forLabel( 0, 1 ), ":Person(age)", 0.5d, 1L );
         visitor.apply( null ).visitUniqueIndex( NewIndexDescriptorFactory.forLabel( 0, 0 ), ":Person(name)", 0.5d, 1L );
-        visitor.apply( null ).visitUniqueConstraint( new UniquenessConstraint( new NodePropertyDescriptor( 1, 0) ), ":Party(name)" );
+        final int[] propertyKeyIds = new int[]{0};
```
Comment (fickludd) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/util/dbstructure/DbStructureInvocationTracingAcceptanceTest.java]: ping

**Review by fickludd — CHANGES_REQUESTED** [review]
This is not really clean yet. Needs another iteration.

File: community/kernel/src/main/java/org/neo4j/kernel/api/schema/EntityPropertyDescriptor.java (hunk 1 above)
Comment (ragadeeshu) [reply community/kernel/src/main/java/org/neo4j/kernel/api/schema/EntityPropertyDescriptor.java:80]: As soon as we remove the RelationshipPropertyDescriptor!

File: community/kernel/src/main/java/org/neo4j/kernel/api/schema_new/constaints/ConstraintBoundary.java (hunk 3 above)
Comment (ragadeeshu) [reply community/kernel/src/main/java/org/neo4j/kernel/api/schema_new/constaints/ConstraintBoundary.java]: If you look closely this is even more stupid, I can simply pass the schema on.

File: community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/IndexProcedures.java (hunk 4 above)
Comment (ragadeeshu) [reply community/kernel/src/main/java/org/neo4j/kernel/builtinprocs/IndexProcedures.java:58]: :+1:

File: community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexIT.java [hunk 19]
```diff
@@ -58,15 +57,17 @@
...
     @Before
     public void createLabelAndProperty() throws Exception
     {
         TokenWriteOperations tokenWrites = tokenWriteOperationsInNewTransaction();
         labelId = tokenWrites.labelGetOrCreateForName( LABEL );
         propertyKeyId = tokenWrites.propertyKeyGetOrCreateForName( PROPERTY_KEY );
-        descriptor = new NodePropertyDescriptor( labelId, propertyKeyId );
+        final int labelId1 = labelId;
+        final int[] propertyKeyIds = new int[]{propertyKeyId};
+        descriptor = SchemaDescriptorFactory.forLabel( labelId1, propertyKeyIds );
```
Comment (fickludd) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/api/index/IndexIT.java]: Another one

File: community/kernel/src/test/java/org/neo4j/kernel/impl/api/integrationtest/UniquenessConstraintValidationIT.java [hunk 20]
```diff
@@ -300,8 +301,10 @@ public void unrelatedNodesWithSamePropertyShouldNotInterfereWithUniquenessCheck(
         ReadOperations readOps = statement.readOperations();
         int person = readOps.labelGetForName( "Person" );
         int propId = readOps.propertyKeyGetForName( "id" );
+        final int labelId = person;
+        final int[] propertyKeyIds = new int[]{propId};
         NewIndexDescriptor idx = readOps
-                .uniqueIndexGetForLabelAndPropertyKey( new NodePropertyDescriptor( person, propId ) );
+                .uniqueIndexGetForLabelAndPropertyKey( SchemaDescriptorFactory.forLabel( labelId, propertyKeyIds ) );
```
Comment (fickludd) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/api/integrationtest/UniquenessConstraintValidationIT.java]: ping

File: community/kernel/src/test/java/org/neo4j/kernel/impl/api/integrationtest/UniquenessConstraintValidationIT.java [hunk 21]
```diff
@@ -327,8 +330,10 @@ public void addingUniqueNodeWithUnrelatedValueShouldNotAffectLookup() throws Exc
...
```
Comment (fickludd) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/api/integrationtest/UniquenessConstraintValidationIT.java]: ping

File: community/kernel/src/test/java/org/neo4j/kernel/impl/util/dbstructure/CineastsDbStructure.java [hunk 22]
```diff
@@ -83,7 +84,8 @@ public void accept( DbStructureVisitor visitor )
         visitor.visitIndex( NewIndexDescriptorFactory.forLabel( 3, 5 ), ":Actor(name)", 1.0d, 44689L );
         visitor.visitIndex( NewIndexDescriptorFactory.forLabel( 4, 5 ), ":Director(name)", 1.0d, 6010L );
         visitor.visitUniqueIndex( NewIndexDescriptorFactory.forLabel( 2, 3 ), ":User(login)", 1.0d, 45L );
-        visitor.visitUniqueConstraint( new UniquenessConstraint( new NodePropertyDescriptor( 2, 3 ) ), "CONSTRAINT ON ( user:User ) ASSERT user.login IS UNIQUE" );
+        final int[] propertyKeyIds = new int[]{3};
+        visitor.visitUniqueConstraint( new UniquenessConstraint( SchemaDescriptorFactory.forLabel( 2, propertyKeyIds ) ), "CONSTRAINT ON ( user:User ) ASSERT user.login IS UNIQUE" );
```
Comment (fickludd) [inline community/kernel/src/test/java/org/neo4j/kernel/impl/util/dbstructure/CineastsDbStructure.java]: pingeling

**Review by fickludd — CHANGES_REQUESTED** [review]
4 more

**Review by fickludd — APPROVED** [review]

**Comment by ragadeeshu** [comment]
@fickludd those HA test pass locally, flaky?

**Comment by fickludd** [comment]
@ragadeeshu yeah, maybe they are ok. Now needs rebase anyway though.

**Pull request merged**